File: frontend/src/services/compra.service.js
import api from './api';
// ── Service ───────────────────────────────────────────────────
export const compraService = {
    async listar(filtros = {}) {
        const params = new URLSearchParams();
        if (filtros.proveedor)
            params.append('proveedor', filtros.proveedor);
        if (filtros.estado)
            params.append('estado', filtros.estado);
        if (filtros.desde)
            params.append('desde', filtros.desde);
        if (filtros.hasta)
            params.append('hasta', filtros.hasta);
        params.append('pagina', String(filtros.pagina ?? 1));
        params.append('tamano', String(filtros.tamano ?? 20));
        const { data } = await api.get(`/compras?${params.toString()}`);
        return data.data;
    },
    async obtener(id) {
        const { data } = await api.get(`/compras/${id}`);
        return data.data;
    },
    async registrar(req) {
        const { data } = await api.post('/compras', req);
        return data.data;
    },
    async aprobar(id) {
        await api.patch(`/compras/${id}/aprobar`);
    },
    async completar(id) {
        await api.patch(`/compras/${id}/completar`);
    },
    async anular(id, motivo) {
        await api.patch(`/compras/${id}/anular`, { motivo });
    },
    async buscarProductos(q) {
        const params = new URLSearchParams();
        if (q)
            params.append('q', q);
        const { data } = await api.get(`/compras/productos/buscar?${params.toString()}`);
        return data.data;
    },
    async buscarProveedores(q) {
        const params = new URLSearchParams();
        if (q)
            params.append('q', q);
        const { data } = await api.get(`/compras/proveedores/buscar?${params.toString()}`);
        return data.data;
    },
    async listarMetodosPago() {
        const { data } = await api.get('/compras/metodos-pago');
        return data.data;
    },
    async buscarRepresentantes(q) {
        const params = new URLSearchParams();
        if (q)
            params.append('q', q);
        const { data } = await api.get(`/compras/representantes/buscar?${params.toString()}`);
        return data.data;
    }
};

File: backend/src/main/java/com/tellix/modules/compra/CompraService.java
package com.tellix.modules.compra;

import com.tellix.security.TellixUserDetails;
import com.tellix.shared.dto.PagedResponse;
import com.tellix.shared.exception.RecursoNoEncontradoException;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
public class CompraService {

    private final CompraRepository repo;

    public CompraService(CompraRepository repo) {
        this.repo = repo;
    }

    public PagedResponse<CompraDto.CompraResumen> listar(
        String proveedor, String estado,
        LocalDate desde, LocalDate hasta,
        int pagina, int tamano
    ) {
        return repo.listar(proveedor, estado, desde, hasta, pagina, tamano);
    }

    public CompraDto.CompraDetalle obtener(int id) {
        return repo.obtener(id)
            .orElseThrow(() -> new RecursoNoEncontradoException("Compra", id));
    }

    public int registrar(CompraDto.CrearCompraRequest req) {
        return repo.registrar(req, usuarioActual().getCodigoUsuario());
    }

    public void anular(int id, CompraDto.AnularCompraRequest req) {
        repo.anular(id, usuarioActual().getCodigoUsuario(),
            req != null ? req.motivo() : null);
    }

    public List<CompraDto.ProductoCompra> buscarProductos(String busqueda) {
        return repo.buscarProductos(busqueda);
    }

    public List<CompraDto.ProveedorCompra> buscarProveedores(String busqueda) {
        return repo.buscarProveedores(busqueda);
    }

    public List<CompraDto.RepresentanteCompra> listarRepresentantes(String proveedor) {
        return repo.listarRepresentantes(proveedor);
    }

    private TellixUserDetails usuarioActual() {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        return (TellixUserDetails) auth.getPrincipal();
    }
}
